File: math/reverse-integer.py
class Solution(object):
    def reverse(self, x):
        """
        :type x: int
        :rtype: int
        """
        # init
        flag = 1

        # check negative number
        if x < 0:
            flag = 0

        # convert to string
        x_str = str(abs(x))

        # reverse the string
        reversed_str = x_str[::-1]

        # convert back to int
        result = int(reversed_str)

        if not flag:
            result = -result

        if result < -2**31 or result > 2**31 - 1:
            return 0

        return result

File: math/reverse-integer.cpp
class Solution {
public:
    int reverse(int x) {

        // use stack

        // init
        stack<int> s;

        string x_str = to_string(x);
        int x_len = x_str.size();

        // check negative number
        bool negative = false;
        int start = 0;

        if (x_str[0] == '-') {
            negative = true;
            start = 1;
        }

        // push all digits into stack
        for (int i = start; i < x_len; i++) {
            s.push(x_str[i] - '0');
        }

        int result = 0;

        // build reversed number
        while (!s.empty()) {
            int digit = s.top();
            s.pop();

            // check overflow before result * 10
            if (result > INT_MAX / 10 ||
                (result == INT_MAX / 10 && digit > 7)) {
                return 0;
            }

            result = result * 10 + digit;
        }

        if (negative) {
            result = -result;
        }

        return result;
    }
};